src/lib/Templates.ml

```diff
@@ -0,0 +1,173 @@
+            let then_branch' = instantiate_instructions param_map then_branch in
+            let else_branch_opt' = Option.map (instantiate_instructions param_map) else_branch_opt in
+            Parsed_if (cond', then_branch', else_branch_opt') :: tl'
+        | Parsed_local_decl _ ->
+            raise (Exceptions.InternalError "[instantiate_instructions]: This point of code is unreachable") 
+
+let instantiate_transition param_map (guard, bloc, sync, loc_name) =
+  let guard' = instantiate_convex_predicate param_map guard in
+  let bloc' = instantiate_instructions param_map bloc in
+  (guard', bloc', sync, loc_name)
+
+let instantiate_transitions param_map =
+  List.map (instantiate_transition param_map)
+
+let instantiate_loc param_map loc =
+  { loc with invariant   = instantiate_convex_predicate param_map loc.invariant;
+             stopped     = instantiate_stopped param_map loc.stopped;
+             flow        = instantiate_flows param_map loc.flow;
+             transitions = instantiate_transitions param_map loc.transitions
+  }
+
+let rename_action_decls param_map user_name action_decls =
+  let rename_action_decl action_name =
+    match Hashtbl.find_opt param_map action_name with
+      | None -> action_name ^ "_" ^ user_name
+      | Some (Arg_name action_name') -> action_name'
+      | Some _ -> failwith "[rename_action_decl]: unexpected argument for template (expecting name)"
+  in
+  List.map rename_action_decl action_decls
+
+let rename_action_locs param_map user_name locs =
+  let rename_action_sync sync =
+    match sync with
+    | NoSync -> NoSync
+    | Sync action_name ->
+        match Hashtbl.find_opt param_map action_name with
+          | None -> Sync (action_name ^ "_" ^ user_name)
+          | Some (Arg_name action_name') -> Sync (action_name')
+          | _ -> failwith "[rename_action_sync]: unexpected argument for template (expecting name)"
+  in
+  let rename_action_transition (guard, bloc, sync, loc_name) =
+    let sync' = rename_action_sync sync in
+    (guard, bloc, sync', loc_name)
+  in
+  let rename_action_transitions loc =
+    { loc with transitions = List.map rename_action_transition loc.transitions }
+  in
+  List.map rename_action_transitions locs
+
+let rename_actions param_map user_name template =
+  let action_decls, locs = template.template_body in
+  let action_decls'      = rename_action_decls param_map user_name action_decls in
+  let locs'              = rename_action_locs param_map user_name locs in
+  { template with template_body = (action_decls', locs') }
+
+(* This is just instantiation, we do type checking somewhere else *)
+let instantiate_automaton templates parsed_template_call =
+    let user_name, template_name, args = parsed_template_call in
+    let template                       = List.find (fun t -> t.template_name = template_name) templates in
+    assert (List.length template.template_parameters = List.length args);
```

Comment:
Could you please show an error message here. I'm not sure that a normal user will understand the error on the terminal 😨

Reply:
good point!